### PR TITLE
Added Discrete Scale to Plot Sub-Dialog 

### LINKED ISSUE #1789: adding discrete scales options to plot sub dialog
By default ggplot ignores unused factor levels, but it should be easy to include them. e.g. for bar chart it can be done like this: http://stackoverflow.com/a/10835659/6424231
We should look into how we could have these options generally.

### PR DESCRIPTION
**Pull request by Vitalis95**
Replaces PR #8395 
Fixes #1789 
@rdstern , @N-thony , @lloyddewit ,the scale is now implemented  only in `Line Plot dialog` , once the PR for Esther on `Bar graphs` is merged then I will implement there. Once it is working well and the code is fine, then we can replicate it in all the graph dialogs

### CODE REVIEW

**Review by rdstern — CHANGES_REQUESTED** [review]
@Vitalis95 I tried with a simple example of 3 climatic stations and months as the starting point.

image

a) It defaul;ts to the x-axis as continuous.  I thought it might notice that the x-axis was a factor and hence discrete?  I thought that's partly why you need some code in the line-plot dialog?  But I am happy to have to move to discrete myself, if that is not the case.
b) Then I used the option 3 to 9 in the limits on the left and that seems to work.  Good.
c) But it keeps coming back to Continuous when I return to the sub-dialog.  It should remember Discrete from the last time.
d) Then I tried Limits on the right, where I assume it must be given as `Mar, Sep`.  Nothing happened.  I can't get anything to happen with the 3 (temporary) controls on the right.  I thought this stop-gap was there to work quickly.  But nothing works there I think.
e) I had closed the limits on the left, but they suddenly returned.  There is something you can't be clearing in the code there.
f) Is there anytrhing else I should be checking?

Please continue, and keep asking @N-thony.  There still seems such a long way to go!

**Comment by lloyddewit** [comment]
> @lloyddewit this is a really important change. It will affect many dialogs, and I wonder whether the code could usefully be examined now, when what is done so far seems to be working ok?

This is a large change but I think I already reviewed a lot of the code in PR #8395. I think it's best if I review when the code is stable an you have approved, otherwise there's a risk I'll miss something.

**Comment by Vitalis95** [comment]
@rdstern , have a look it , line graphs and general graphics dialogs. I will be making small change of just adding the discrete function in other graphs dialog where discrete applies

**Review by rdstern — CHANGES_REQUESTED** [review]
@Vitalis95 there are 3 types of point here:
a) Great that you have added the other discrete options.  Could you please be exact on the spacing and position of the new checkbox.  Also you have included the secondary axis.  I don't think that applies to the discrete option.  (Perhaps I am wrong, but if you thought I had suggested it, then please remove it.)
b) It does seem to work from the general dialog, but only once, because the sub-dialog doesn't keep any memory of having used those options.  So it returns immedaitely to the defaults.
c) I thought that discussions with @Patowhiz had confirmed that nothing was needed on each main dialog.  Did I get that impression wrongly?

File: instat/ucrAxes.vb [hunk 1]
```diff
@@ -359,13 +469,48 @@ Public Class ucrAxes
...
         clsDummyFunction = New RFunction
         clsDummyFunction.AddParameter("name", "auto", iPosition:=1)
+        clsDummyFunction.AddParameter("long_label", "abbr", iPosition:=2)
 
         clsXYSecondaryAxisFunction = New RFunction
         clsXYSecondaryAxisFunction.SetRCommand("sec_axis")
 
-        clsXYlabTitleFunction = clsNewXYlabTitleFunction
-        clsXYScaleDateFunction = clsNewXYScaleDateFunction
-        clsXYScaleContinuousFunction = clsNewXYScaleContinuousFunction
```
[the remark this answers is not in the document]
Comment (Vitalis95) [reply instat/ucrAxes.vb:368]: it was moved up

**Comment by Vitalis95** [comment]
@Patowhiz , are we suppose to add the discrete function in the main graphs dialog, in the sdgplots sub , like the way they  have added for continuous scale function and date function?

**Review by rdstern — COMMENTED** [review]
@Vitalis95 I am now ready to see this merged, as soon as it is available for more of the graphics dialogs.
@Patowhiz and @N-thony I didn't realise there is already a function in each main graph dialog for continuous and date?  If so, the clearly it could usefully be amended for discrete.  I am still unclear whether it is a different function in each main graph dialog, or a single function to adapt that is called by each main graph dialog?

**Comment by Patowhiz** [comment]
@Vitalis95 I think we discussed about this feature in skype. I'm happy to expound on it in an online discussion. 

I noticed there are features that were added in main dialogs for this subdialog in a way that introduced more unnecessary parameters when setting up the subdialog. I think we should be keen on avoiding that trend.

Thanks.

**Review by rdstern — COMMENTED** [review]
@Vitalis95 and @N-thony well done both of you.  I tried with the bar charts and that seems to work, also with the scatterplots.  This is on the home run now.

Could you just check the Mosaic plot code, see below:

image

This is a good candidate, to label better but when I try it says the x-scale is continuous - which it isn't!

**Comment by Vitalis95** [comment]
@N-thony , the get  parameter value from clsRaesFunction works when the main dialog has a single receiver, for multiple receivers like for dlgMossaic plot  raised above or when you switch the receiver in dlgBargraph to multiple receiver, it  doesn't get x variables

**Comment by rdstern** [comment]
@Vitalis95 that's a good bit of detective work.  I hope it is therefore an easy fix for @N-thony

**Comment by Vitalis95** [comment]
> > @N-thony , the get parameter value from clsRaesFunction works when the main dialog has a single receiver, for multiple receivers like for dlgMossaic plot raised above or when you switch the receiver in dlgBargraph to multiple receiver, it doesn't get x variables
> 
> @Vitalis95 what do you think? Have you tried to run on Script window or check the documentation if it is possible when `x` has more than one variable?

@lilyclements is looking on the R-code for this.

**Comment by rdstern** [comment]
@Vitalis95 I just wondered if it was ready now, though perhaps you are still waiting for @lilyclements?  However, it has conflicts now, so doesn't build.

**Comment by lilyclements** [comment]
@Vitalis95 I've sent you the code for this now :)

**Comment by Vitalis95** [comment]
@lilyclements , I get a bug when I run the code on  dlg mosaic plot, check your mattermost I have sent the code there.

**Comment by lilyclements** [comment]
@Vitalis95 great observation - very indepth to be looking at how this affects other dialogs too. From a little google, it seems [this is a known bug in `ggmosaic`](https://github.com/haleyjeppson/ggmosaic/issues/58).
We should keep an eye on it, but, perhaps for now is it possible to disable that tab if you access the sub dialog from a mosaic plot? If not, then I'm sure we can catch it in another way.

**Comment by rdstern** [comment]
@Vitalis95 I now get conflicts so can't check this pull request any more.  If you can resolve them, then I'd also like to test the pull request even before you do anything more.  From @lilyclements if this is a sort of bug in the mosaicplot package, then, for now, I would accept an error when using it in R-Instat.  The mosaic plot is not important enough to prevent adding this important new facility.  And we need to get on, with important new changes.

**Comment by lilyclements** [comment]
@rdstern yes this is the case. In which instance, @Vitalis95 no need to worry about the mosaic bug.

**Comment by lilyclements** [comment]
@rdstern @Vitalis95 the conflict should now be fixed

**Comment by rdstern** [comment]
@lilyclements I still get build errors.  I have updated the master to the latest version and am not clear what else I should do?

**Comment by Vitalis95** [comment]
@rdstern @N-thony , the merge conflicts is because of Derrick's  PR #8455 , we also made a change on dlg bar graph. Once this is merged I will make another PR.

**Comment by Vitalis95** [comment]
merge conflicts

**Pull request closed without merging**